chain/indexer/tipset.go

```diff
@@ -0,0 +1,408 @@
+	"github.com/filecoin-project/lily/chain/indexer/processor"
+	"github.com/filecoin-project/lily/metrics"
+	"github.com/filecoin-project/lily/model"
+	visormodel "github.com/filecoin-project/lily/model/visor"
+	"github.com/filecoin-project/lily/tasks"
+	"github.com/filecoin-project/lily/tasks/actorstate"
+	init_2 "github.com/filecoin-project/lily/tasks/actorstate/init_"
+	market2 "github.com/filecoin-project/lily/tasks/actorstate/market"
+	miner2 "github.com/filecoin-project/lily/tasks/actorstate/miner"
+	multisig2 "github.com/filecoin-project/lily/tasks/actorstate/multisig"
+	power2 "github.com/filecoin-project/lily/tasks/actorstate/power"
+	"github.com/filecoin-project/lily/tasks/actorstate/raw"
+	reward2 "github.com/filecoin-project/lily/tasks/actorstate/reward"
+	verifreg2 "github.com/filecoin-project/lily/tasks/actorstate/verifreg"
+	"github.com/filecoin-project/lily/tasks/blocks"
+	"github.com/filecoin-project/lily/tasks/chaineconomics"
+	"github.com/filecoin-project/lily/tasks/consensus"
+	"github.com/filecoin-project/lily/tasks/indexer"
+	"github.com/filecoin-project/lily/tasks/messageexecutions/internal_message"
+	"github.com/filecoin-project/lily/tasks/messageexecutions/internal_parsed_message"
+	"github.com/filecoin-project/lily/tasks/messages/block_message"
+	"github.com/filecoin-project/lily/tasks/messages/gas_economy"
+	"github.com/filecoin-project/lily/tasks/messages/gas_output"
+	"github.com/filecoin-project/lily/tasks/messages/message"
+	"github.com/filecoin-project/lily/tasks/messages/parsed_message"
+	"github.com/filecoin-project/lily/tasks/messages/receipt"
+	"github.com/filecoin-project/lily/tasks/msapprovals"
+)
+
+var tsLog = logging.Logger("lily/index/tipset")
+
+// TipSetIndexer waits for tipsets and persists their block data into a database.
+type TipSetIndexer struct {
+	name  string
+	node  tasks.DataSource
+	tasks []string
+
+	procBuilder *processor.Builder
+}
+
+type TipSetIndexerOpt func(t *TipSetIndexer)
+
+// NewTipSetIndexer extracts block, message and actor state data from a tipset and persists it to storage. Extraction
+// and persistence are concurrent. Extraction of the a tipset can proceed while data from the previous extraction is
+// being persisted. The indexer may be given a time window in which to complete data extraction. The name of the
+// indexer is used as the reporter in the visor_processing_reports table.
+func NewTipSetIndexer(node tasks.DataSource, name string, tasks []string, options ...TipSetIndexerOpt) (*TipSetIndexer, error) {
+	tsi := &TipSetIndexer{
+		name:  name,
+		node:  node,
+		tasks: tasks,
+	}
+
+	tipsetProcessors := map[string]processor.TipSetProcessor{}
+	tipsetsProcessors := map[string]processor.TipSetsProcessor{}
+	actorProcessors := map[string]processor.ActorProcessor{}
+	reportProcessors := map[string]processor.ReportProcessor{
+		"builtin": indexer.NewTask(node),
+	}
+
+	for _, t := range tasks {
+		switch t {
```

Comment:
Seeking ideas on how to make this switch less verbose.

Reply:
Not to throw interfaces at this again, but if all initialization could be done behind an interface w something like `Init(<all deps over all tasks>...)` they could also be done in a forloop. Each Init doesn't need to use all of the args.